File: sdk/src/main/java/eu/darken/porter/sdk/PorterUserServiceCodec.java
package eu.darken.porter.sdk;

import static eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_COMPONENT;
import static eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_DAEMON;
import static eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_DEBUGGABLE;
import static eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_NO_CREATE;
import static eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_PROCESS_NAME_SUFFIX;
import static eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_REMOVE;
import static eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_TAG;
import static eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_USE_32_BIT;
import static eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_VERSION_CODE;

import android.os.Bundle;

import androidx.annotation.NonNull;

import java.util.Objects;

/** How {@link Porter.UserServiceArgs} is written in the Porter server's key space. */
final class PorterUserServiceCodec {

    private PorterUserServiceCodec() {
    }

    /** @param noCreate marks the service "do not start it", as {@code peekUserService} asks for */
    @NonNull
    static Bundle encodeUserService(@NonNull Porter.UserServiceArgs args, boolean noCreate) {
        Bundle bundle = new Bundle();
        bundle.putParcelable(USER_SERVICE_COMPONENT, args.componentName);
        bundle.putBoolean(USER_SERVICE_DEBUGGABLE, args.debuggable);
        bundle.putInt(USER_SERVICE_VERSION_CODE, args.versionCode);
        bundle.putBoolean(USER_SERVICE_DAEMON, args.daemon);
        bundle.putBoolean(USER_SERVICE_USE_32_BIT, args.use32BitAppProcess);
        bundle.putString(USER_SERVICE_PROCESS_NAME_SUFFIX,
                Objects.requireNonNull(args.processName, "process name suffix must not be null"));
        if (args.tag != null) {
            bundle.putString(USER_SERVICE_TAG, args.tag);
        }
        if (noCreate) {
            bundle.putBoolean(USER_SERVICE_NO_CREATE, true);
        }
        return bundle;
    }

    @NonNull
    static Bundle encodeUserServiceRemoval(@NonNull Porter.UserServiceArgs args, boolean remove) {
        Bundle bundle = new Bundle();
        bundle.putParcelable(USER_SERVICE_COMPONENT, args.componentName);
        if (args.tag != null) {
            bundle.putString(USER_SERVICE_TAG, args.tag);
        }
        bundle.putBoolean(USER_SERVICE_REMOVE, remove);
        return bundle;
    }
}

File: sdk/src/main/java/eu/darken/porter/sdk/PorterUserServiceCodec.kt
package eu.darken.porter.sdk

import android.os.Bundle
import eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_COMPONENT
import eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_DAEMON
import eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_DEBUGGABLE
import eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_NO_CREATE
import eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_PROCESS_NAME_SUFFIX
import eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_REMOVE
import eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_TAG
import eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_USE_32_BIT
import eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_VERSION_CODE

/**
 * How [UserServiceArgs] is written in the Porter server's key space. Reachable from the server's
 * own tests, which decode what this writes.
 */
internal object PorterUserServiceCodec {

    /** @param noCreate marks the service "do not start it", as a peek asks for */
    fun encodeUserService(args: UserServiceArgs, noCreate: Boolean): Bundle = Bundle().apply {
        putParcelable(USER_SERVICE_COMPONENT, args.componentName)
        putBoolean(USER_SERVICE_DEBUGGABLE, args.debuggable)
        putInt(USER_SERVICE_VERSION_CODE, args.version)
        putBoolean(USER_SERVICE_DAEMON, args.daemon)
        putBoolean(USER_SERVICE_USE_32_BIT, false)
        putString(USER_SERVICE_PROCESS_NAME_SUFFIX, args.processNameSuffix)
        if (args.tag != null) {
            putString(USER_SERVICE_TAG, args.tag)
        }
        if (noCreate) {
            putBoolean(USER_SERVICE_NO_CREATE, true)
        }
    }

    fun encodeUserServiceRemoval(args: UserServiceArgs, remove: Boolean): Bundle = Bundle().apply {
        putParcelable(USER_SERVICE_COMPONENT, args.componentName)
        if (args.tag != null) {
            putString(USER_SERVICE_TAG, args.tag)
        }
        putBoolean(USER_SERVICE_REMOVE, remove)
    }
}
